setup.py

```diff
@@ -258,86 +187,96 @@ def run(self):
-    'bazel-xla/external/com_google_protobuf/src',
-    'bazel-xla/external/eigen_archive',
-    'bazel-xla/external/com_google_absl',
-    'bazel-xla/external/com_googlesource_code_re2',
-    'bazel-xla/com_github_grpc_grpc/include',
-]:
-  include_dirs.append(os.path.join(base_dir, ipath))
-include_dirs += [
-    pytorch_source_path,
-    os.path.join(pytorch_source_path, 'torch/csrc'),
-    os.path.join(pytorch_source_path, 'torch/lib/tmp_install/include'),
-]
-
-library_dirs = []
-library_dirs.append(lib_path)
-
-extra_link_args = []
-
 DEBUG = _check_env_flag('DEBUG')
 IS_DARWIN = (platform.system() == 'Darwin')
+IS_WINDOWS = sys.platform.startswith('win')
 IS_LINUX = (platform.system() == 'Linux')
+GCLOUD_KEY_FILE = os.getenv('GCLOUD_SERVICE_KEY_FILE', default='')
+CACHE_SILO_NAME = os.getenv('SILO_NAME', default='dev')
+BAZEL_JOBS = os.getenv('BAZEL_JOBS', default='')
 
+extra_compile_args = []
+cxx_abi = getattr(torch._C, '_GLIBCXX_USE_CXX11_ABI', None)
+if cxx_abi is not None:
+  extra_compile_args += ['-D_GLIBCXX_USE_CXX11_ABI={}'.format(int(cxx_abi))]
 
-def make_relative_rpath(path):
-  if IS_DARWIN:
-    return '-Wl,-rpath,@loader_path/' + path
-  else:
-    return '-Wl,-rpath,$ORIGIN/' + path
 
+class BazelExtension(Extension):
+  """A C/C++ extension that is defined as a Bazel BUILD target."""
+
+  def __init__(self, bazel_target):
+    self.bazel_target = bazel_target
+    self.relpath, self.target_name = (
+        posixpath.relpath(bazel_target, '//').split(':'))
+    ext_name = os.path.join(
+        self.relpath.replace(posixpath.sep, os.path.sep), self.target_name)
+    if ext_name.endswith('.so'):
+      ext_name = ext_name[:-3]
+    Extension.__init__(self, ext_name, sources=[])
 
-extra_compile_args = [
-    '-std=c++17',
-    '-Wno-sign-compare',
-    '-Wno-deprecated-declarations',
-    '-Wno-return-type',
-]
 
-if re.match(r'clang', os.getenv('CC', '')):
-  extra_compile_args += [
-      '-Wno-macro-redefined',
-      '-Wno-return-std-move',
-  ]
+class BuildBazelExtension(command.build_ext.build_ext):
+  """A command that runs Bazel to build a C/C++ extension."""
 
-if DEBUG:
-  extra_compile_args += ['-O0', '-g']
-  extra_link_args += ['-O0', '-g']
-else:
-  extra_compile_args += ['-DNDEBUG']
+  def run(self):
+    for ext in self.extensions:
+      self.bazel_build(ext)
+    command.build_ext.build_ext.run(self)
+
+  def bazel_build(self, ext):
+    if not os.path.exists(self.build_temp):
+      os.makedirs(self.build_temp)
+
+    bazel_argv = [
+        'bazel', 'build', ext.bazel_target,
+        f"--symlink_prefix={os.path.join(self.build_temp, 'bazel-')}",
+        '\n'.join(['--cxxopt=%s' % opt for opt in extra_compile_args])
+    ]
+
+    # Debug build.
+    if DEBUG:
+      bazel_argv.append('--compilation_mode=dbg')
+
+    # Remote cache authentication.
+
+    if _check_env_flag('BAZEL_REMOTE_CACHE'):
+      bazel_argv.append('--config=remote_cache')
+
+    if GCLOUD_KEY_FILE:
```

Comment:
can you explain to me why we need gcloud during build?

Reply:
We're using gcloud remote build cache on gcloud and for that we need to authenticate with it.

There are 3 scenarions

 - gcloud/workstation will use default authentication
 - circleci will use the service account key
 - cloudbuild will use the cloudbuild service account authentication

https://developers.google.com/remote-build-execution/docs/remote-cache-overview